# Содержание

1. Цель исследования
2. Подключение к базе данных и необходимых библиотек
   - Импортируем библиотеки.
   - Установим доступ к базе данных.
3. Ознакомимся с данными таблиц
   - Выведем первые пять строк таблиц.
4. Посчитаем, сколько книг вышло после 1 января 2000 года
5. Для каждой книги посчитаем количество обзоров и среднюю оценку
6. Определим издательство, которое выпустило наибольшее число книг толще 50 страниц 
7. Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками
8. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок
9. Вывод

## Цель исследования

Сформулировать ценностное предложение для сервиса чтения книг.

Цель: 
Анализ базы данных крупного книжного сервиса для чтения книг по подписке. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

Задачи:
Ознакомиться с данными таблиц
Посчитать, сколько книг вышло после 1 января 2000 года
Для каждой книги посчитать количество обзоров и среднюю оценку
Определить издательство, которое выпустило наибольшее число книг толще 50 страниц
Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками
Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок

## Подключение к базе данных и необходимых библиотек

### Импортируем библиотеки

In [1]:
import pandas as pd
from sqlalchemy import create_engine

### Установим доступ к базе данных

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Ознакомимся с данными таблиц

### Выведем первые пять строк таблиц

In [3]:
sql_query = """
  SELECT *
  FROM books, authors, publishers, ratings, reviews
  LIMIT 5;
  """

In [4]:
tables = pd.io.sql.read_sql(sql_query, con = engine)
tables

book_id  author_id         title  num_pages publication_date  publisher_id  \
0        1        546  'Salem's Lot        594       2005-11-01            93   
1        1        546  'Salem's Lot        594       2005-11-01            93   
2        1        546  'Salem's Lot        594       2005-11-01            93   
3        1        546  'Salem's Lot        594       2005-11-01            93   
4        1        546  'Salem's Lot        594       2005-11-01            93   

   author_id      author  publisher_id publisher  rating_id  book_id  \
0          1  A.S. Byatt             1       Ace          1        1   
1          1  A.S. Byatt             1       Ace          2        1   
2          1  A.S. Byatt             1       Ace          3        1   
3          1  A.S. Byatt             1       Ace          4        2   
4          1  A.S. Byatt             1       Ace          5        2   

        username  rating  review_id  book_id      username  \
0     ryanfranco       4          1        1  brandtandrea   
1  grantpatricia       2          1        1  brandtandrea   
2   brandtandrea       5          1        1  brandtandrea   
3       lorichen       3          1        1  brandtandrea   
4    mariokeller       2          1        1  brandtandrea   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Mention society tell send professor analysis. ...  
2  Mention society tell send professor analysis. ...  
3  Mention society tell send professor analysis. ...  
4  Mention society tell send professor analysis. ...

Столбцы всех 5-ти таблиц в соответствии с схемой базы данных присутствуют.

In [5]:
display(tables.info(), tables.isna().sum(), tables.duplicated().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           5 non-null      int64 
 1   author_id         5 non-null      int64 
 2   title             5 non-null      object
 3   num_pages         5 non-null      int64 
 4   publication_date  5 non-null      object
 5   publisher_id      5 non-null      int64 
 6   author_id         5 non-null      int64 
 7   author            5 non-null      object
 8   publisher_id      5 non-null      int64 
 9   publisher         5 non-null      object
 10  rating_id         5 non-null      int64 
 11  book_id           5 non-null      int64 
 12  username          5 non-null      object
 13  rating            5 non-null      int64 
 14  review_id         5 non-null      int64 
 15  book_id           5 non-null      int64 
 16  username          5 non-null      object
 17  text              5 

None

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
author_id           0
author              0
publisher_id        0
publisher           0
rating_id           0
book_id             0
username            0
rating              0
review_id           0
book_id             0
username            0
text                0
dtype: int64

0

Данные чистые пропусков и дубликатов нет, тип данных неверный только у столбца с датой 'publication_date'.

## Посчитаем, сколько книг вышло после 1 января 2000 года

In [6]:
sql_query_one = """
  SELECT COUNT(DISTINCT(book_id))
  FROM books
  WHERE publication_date > '2000-01-01';
  """

In [7]:
pd.io.sql.read_sql(sql_query_one, con = engine)

count
0    819

С начала 2000 года вышло 819 книг.

## Для каждой книги посчитаем количество обзоров и среднюю оценку

In [8]:
sql_query_two = """
  WITH
  rew AS (SELECT book_id, review_id
          FROM reviews),
          
  rat AS (SELECT rating_id, rating
          FROM ratings),
          
  bk AS (SELECT DISTINCT book_id, title
         FROM books)
  
  SELECT COUNT(review_id), AVG(rating), bk.book_id
  FROM bk
  LEFT OUTER JOIN rew ON bk.book_id = rew.book_id
  LEFT OUTER JOIN rat ON bk.book_id = rat.rating_id
  GROUP BY bk.book_id;
  """

In [9]:
pd.io.sql.read_sql(sql_query_two, con = engine)

count  avg  book_id
0        2  4.0        1
1        1  2.0        2
2        3  5.0        3
3        2  3.0        4
4        4  2.0        5
..     ...  ...      ...
995      3  4.0      996
996      3  5.0      997
997      4  5.0      998
998      2  5.0      999
999      4  3.0     1000

[1000 rows x 3 columns]

Все в порядке, данные получены.

## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц 

In [10]:
sql_query_three = """
  WITH         
  bk AS (SELECT book_id, publisher_id, num_pages
         FROM books
         WHERE num_pages > 50),
         
  pb AS (SELECT publisher_id, publisher
         FROM publishers)       
  
  SELECT pb.publisher
  FROM bk
  JOIN pb ON bk.publisher_id = pb.publisher_id
  
  GROUP BY pb.publisher
  ORDER BY COUNT(bk.book_id) DESC
  LIMIT 3;
  """

In [11]:
pd.io.sql.read_sql(sql_query_three, con = engine)

publisher
0             Penguin Books
1                   Vintage
2  Grand Central Publishing

Победило 'Penguin Books', именно оно выпускает наибольшее количество книг с 50+ числом страниц, на 2-ом месте 'Vintage' и на 3-ем 'Grand Central Publishing'.

## Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками

In [12]:
sql_query_four = """
  WITH         
  bk AS (SELECT book_id, author_id, title
           FROM books),
         
  rt AS (SELECT rating_id, rating, book_id
         FROM ratings),
         
  at AS (SELECT author_id, author
         FROM authors)      
  
  SELECT author, AVG(i.rat)
  FROM (SELECT rt.book_id, AVG(rt.rating) AS rat, bk.author_id
        FROM rt
        JOIN bk ON rt.book_id = bk.book_id
        GROUP BY rt.book_id, bk.author_id
        HAVING COUNT(rt.rating_id) >= 50
  ) AS i     
  JOIN at ON i.author_id = at.author_id    
  GROUP BY author
  ORDER BY AVG(i.rat) DESC
  --LIMIT 2;
  """

In [13]:
pd.io.sql.read_sql(sql_query_four, con = engine)

author       avg
0                          J.K. Rowling/Mary GrandPré  4.283844
1                   Markus Zusak/Cao Xuân Việt Khương  4.264151
2                                      J.R.R. Tolkien  4.258446
3                                   Louisa May Alcott  4.192308
4                                        Rick Riordan  4.080645
5                                     William Golding  3.901408
6                                       J.D. Salinger  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
9                                           Dan Brown  3.754540
10                                         Lois Lowry  3.750000
11            George Orwell/Boris Grabnar/Peter Škerl  3.729730
12                                    Stephenie Meyer  3.662500
13                                     John Steinbeck  3.622951

Дж.К.Роулинг с вселенной Гарри Поттер оказалась на вершине популярности, разделив ее с 'Mary GrandPré', Толкин с вселенной Властелина колец занял 3-ое место, на 2-ом Markus Zusak/Cao Xuân Việt Khương.

## Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [14]:
sql_query_five = """

  WITH         
  rt AS (SELECT rating_id, username
         FROM ratings),
         
  rw AS (SELECT review_id, username
         FROM reviews)      
  
  SELECT AVG(i.revi)
  FROM (SELECT COUNT(rw.review_id) AS revi
        FROM rw
        WHERE 
           rw.username IN (SELECT username
                        FROM rt
                        GROUP BY username 
                        HAVING COUNT(rt.rating_id) > 50)
        GROUP BY rw.username) AS i;
  """

In [15]:
pd.io.sql.read_sql(sql_query_five, con = engine)

avg
0  24.333333

Среднее количество обзоров активных пользователей около 24 ед.

## Вывод

В среднем активные пользователи оставляют по 24 обзора
В пятерку лучших авторов вошли:
                              1. J.K. Rowling/Mary GrandPré
                              2. Markus Zusak/Cao Xuân Việt Khương
                              3. J.R.R. Tolkien
                              4. Louisa May Alcott
                              5. Rick Riordan
Самыми продуктивными издательствами являются: 
                                            1. Penguin Books
                                            2. Vintage
                                            3. Grand Central Publishing